File: frontend/src/pipeline.ts
import * as App from '../wailsjs/go/appapi/API'
import { store, pipeline } from '../wailsjs/go/models'

const $ = (id: string) => document.getElementById(id) as HTMLElement

const PATHWAYS = [
  { key: '', label: '— unrouted —' },
  { key: 'rapid_brainstorm', label: 'Rapid Brainstorm' },
  { key: 'side_business', label: 'Side Business' },
  { key: 'small_project', label: 'Small Project' },
  { key: 'tech_product', label: 'Technology / Product-Led' },
  { key: 'full_startup', label: 'Full Startup' },
]

const STATUSES = ['raw', 'triaged', 'in_review', 'validating', 'go', 'parked', 'killed']

let ideas: store.Idea[] = []
let selectedId: string | null = null

function fmtDate(ms: number): string {
  if (!ms) return ''
  const d = new Date(ms)
  return d.toISOString().slice(0, 10)
}

function pathwayLabel(key: string): string {
  return PATHWAYS.find(p => p.key === key)?.label ?? key
}

async function loadIdeas() {
  try {
    ideas = (await App.ListIdeas()) || []
  } catch (e: any) {
    $('ideaList').innerHTML = `<p class="pl-error">Could not load ideas: ${e?.userMessage || e}</p>`
    return
  }
  renderIdeaList()
}

function renderIdeaList() {
  const host = $('ideaList')
  host.innerHTML = ''
  if (ideas.length === 0) {
    host.innerHTML = '<p class="pl-empty">No ideas yet. Add one to start the pipeline.</p>'
    $('ideaDetail').innerHTML = ''
    return
  }
  for (const idea of ideas) {
    const row = document.createElement('div')
    row.className = 'idea-row' + (idea.ID === selectedId ? ' selected' : '')
    row.innerHTML =
      `<span class="idea-title">${idea.Title}</span>` +
      `<span class="status-chip status-${idea.Status}">${idea.Status}</span>` +
      `<span class="idea-pathway">${pathwayLabel(idea.Pathway)}</span>` +
      (idea.FinancialFlag ? '<span class="fin-flag" title="Touches financial data">$</span>' : '')
    row.onclick = () => { selectedId = idea.ID; renderIdeaList(); void renderDetail(idea.ID) }
    host.appendChild(row)
  }
}

async function renderDetail(id: string) {
  const idea = ideas.find(i => i.ID === id)
  if (!idea) return
  const detail = $('ideaDetail')

  let crits: store.KillCriterion[] = []
  try { crits = (await App.ListKillCriteria(id)) || [] } catch { /* show empty */ }

  const statusOptions = STATUSES.filter(s => s !== idea.Status)
    .map(s => `<option value="${s}">${s}</option>`).join('')

  detail.innerHTML = `
    <h2>${idea.Title}</h2>
    <p class="idea-summary">${idea.Summary || '<em>No summary.</em>'}</p>
    <div class="detail-row"><label>Pathway</label> ${pathwayLabel(idea.Pathway)}</div>
    <div class="detail-row"><label>Status</label>
      <span class="status-chip status-${idea.Status}">${idea.Status}</span>
      <select id="statusSel"><option value="">Move to…</option>${statusOptions}</select>
    </div>
    <h3>Kill criteria</h3>
    <table class="kc-table"><thead><tr>
      <th>Metric</th><th>Threshold</th><th>Review date</th><th>On miss</th><th></th>
    </tr></thead><tbody id="kcBody">
      ${crits.map(k => `<tr>
        <td>${k.Metric}</td><td>${k.Threshold}</td>
        <td>${fmtDate(k.ReviewDate)}</td><td>${k.OnMiss}</td>
        <td><button class="kc-del" data-id="${k.ID}">✕</button></td></tr>`).join('')}
    </tbody></table>
    <button id="addKcBtn">+ Add kill criterion</button>
  `

  ;($('statusSel') as HTMLSelectElement).onchange = (e) => {
    const to = (e.target as HTMLSelectElement).value
    if (to) void moveStatus(idea.ID, to)
  }
  $('addKcBtn').onclick = () => void addCriterion(idea.ID)
  detail.querySelectorAll('.kc-del').forEach(btn => {
    ;(btn as HTMLElement).onclick = async () => {
      await App.DeleteKillCriterion((btn as HTMLElement).dataset.id!)
      void renderDetail(idea.ID)
    }
  })
}

async function moveStatus(id: string, to: string) {
  let reason = ''
  if (to === 'killed' || to === 'parked') {
    reason = prompt(`Reason for moving to ${to}:`) || ''
    if (!reason) return
  }
  try {
    await App.SetIdeaStatus(id, to, reason)
  } catch (e: any) {
    alert(e?.userMessage || `Could not change status: ${e}`)
    return
  }
  await loadIdeas()
  void renderDetail(id)
}

async function addCriterion(ideaID: string) {
  const metric = prompt('Metric (e.g. "Paid installs"):')
  if (!metric) return
  const threshold = prompt('Threshold (e.g. ">=2 in 30 days"):')
  if (!threshold) return
  const dateStr = prompt('Review date (YYYY-MM-DD):')
  if (!dateStr) return
  const reviewDate = Date.parse(dateStr + 'T00:00:00Z')
  if (isNaN(reviewDate)) { alert('Invalid date.'); return }
  const onMiss = (prompt('On miss — kill, park, or halt:', 'kill') || '').trim()
  try {
    await App.AddKillCriterion(ideaID, metric, threshold, reviewDate, onMiss)
  } catch (e: any) {
    alert(e?.userMessage || `Could not add criterion: ${e}`)
    return
  }
  void renderDetail(ideaID)
}

async function newIdea() {
  const title = prompt('Idea title:')
  if (!title) return
  const summary = prompt('One-line summary (optional):') || ''
  const pathway = (prompt('Pathway key (side_business, small_project, full_startup, …) or blank:') || '').trim()
  const financial = confirm('Does this idea touch financial data? OK = yes.')
  try {
    const created = await App.CreateIdea(title, summary, pathway, financial)
    selectedId = created.ID
  } catch (e: any) {
    alert(e?.userMessage || `Could not create idea: ${e}`)
    return
  }
  await loadIdeas()
  if (selectedId) void renderDetail(selectedId)
}

export function openPipeline() {
  $('pipelineView').classList.remove('hidden')
  void loadIdeas()
  void refreshReviewsDue()
}

function closePipeline() {
  $('pipelineView').classList.add('hidden')
}

// refreshReviewsDue runs the sweep and updates the sidebar badge + panel.
export async function refreshReviewsDue() {
  let due: pipeline.DueReviewView[] = []
  try { due = (await App.ListReviewsDue()) || [] } catch { return }
  const badge = $('reviewsDueBadge')
  if (due.length > 0) {
    badge.textContent = String(due.length)
    badge.classList.remove('hidden')
  } else {
    badge.classList.add('hidden')
  }
  const panel = $('reviewsDuePanel')
  if (due.length === 0) {
    panel.classList.add('hidden')
    panel.innerHTML = ''
    return
  }
  panel.classList.remove('hidden')
  panel.innerHTML =
    `<div class="rd-title">⏰ ${due.length} review${due.length > 1 ? 's' : ''} due</div>` +
    due.map(d => `<div class="rd-row">
      <strong>${d.IdeaTitle}</strong> — ${d.Metric} (${d.Threshold}),
      due ${fmtDate(d.ReviewDate)}${d.DaysOverdue > 0 ? `, ${d.DaysOverdue}d overdue` : ''}
      → on miss: ${d.OnMiss}</div>`).join('')
}

export function initPipeline() {
  $('pipelineBtn').onclick = () => openPipeline()
  $('pipelineBack').onclick = () => closePipeline()
  $('pipelineNewBtn').onclick = () => void newIdea()
}

File: internal/appapi/pipeline.go
package appapi

import (
	"errors"
	"time"

	"github.com/cajundata/starshp_app/internal/pipeline"
	"github.com/cajundata/starshp_app/internal/provider"
	"github.com/cajundata/starshp_app/internal/store"
	"github.com/google/uuid"
)

// CreateIdea creates a new idea in the 'raw' status and returns it.
func (a *API) CreateIdea(title, summary, pathway string, financialFlag bool) (store.Idea, error) {
	if title == "" {
		return store.Idea{}, provider.AppError{
			Code: "invalid_input", UserMessage: "An idea needs a title.", Retryable: false}
	}
	idea := store.Idea{
		ID: uuid.NewString(), Title: title, Summary: summary, Pathway: pathway,
		Status: "raw", FinancialFlag: financialFlag, Source: "manual",
	}
	if err := a.st.CreateIdea(idea); err != nil {
		return store.Idea{}, err
	}
	return a.st.GetIdea(idea.ID)
}

func (a *API) UpdateIdea(i store.Idea) error    { return a.st.UpdateIdea(i) }
func (a *API) ListIdeas() ([]store.Idea, error) { return a.st.ListIdeas() }
func (a *API) GetIdea(id string) (store.Idea, error) {
	return a.st.GetIdea(id)
}
func (a *API) DeleteIdea(id string) error { return a.st.DeleteIdea(id) }

// SetIdeaStatus validates the transition (legality + reason-required for
// terminal statuses) before persisting it.
func (a *API) SetIdeaStatus(id, toStatus, reason string) error {
	cur, err := a.st.GetIdea(id)
	if err != nil {
		return provider.AppError{Code: "not_found",
			UserMessage: "That idea no longer exists.", Retryable: false}
	}
	if verr := pipeline.ValidateTransition(cur.Status, toStatus, reason); verr != nil {
		var te *pipeline.TransitionError
		if errors.As(verr, &te) {
			return provider.AppError{Code: te.Code, UserMessage: te.Message, Retryable: false}
		}
		return verr
	}
	return a.st.SetIdeaStatus(id, toStatus, reason)
}

func (a *API) ListStatusHistory(ideaID string) ([]store.StatusChange, error) {
	return a.st.ListStatusHistory(ideaID)
}

// AddKillCriterion stores a new kill criterion (status 'pending') and returns it.
func (a *API) AddKillCriterion(ideaID, metric, threshold string, reviewDate int64, onMiss string) (store.KillCriterion, error) {
	if metric == "" || threshold == "" {
		return store.KillCriterion{}, provider.AppError{Code: "invalid_input",
			UserMessage: "A kill criterion needs a metric and a threshold.", Retryable: false}
	}
	switch onMiss {
	case "kill", "park", "halt":
	default:
		return store.KillCriterion{}, provider.AppError{Code: "invalid_input",
			UserMessage: "On-miss must be kill, park, or halt.", Retryable: false}
	}
	k := store.KillCriterion{
		ID: uuid.NewString(), IdeaID: ideaID, Metric: metric, Threshold: threshold,
		ReviewDate: reviewDate, OnMiss: onMiss, Status: "pending",
	}
	if err := a.st.AddKillCriterion(k); err != nil {
		return store.KillCriterion{}, err
	}
	return k, nil
}

func (a *API) UpdateKillCriterion(k store.KillCriterion) error { return a.st.UpdateKillCriterion(k) }
func (a *API) DeleteKillCriterion(id string) error             { return a.st.DeleteKillCriterion(id) }
func (a *API) ListKillCriteria(ideaID string) ([]store.KillCriterion, error) {
	return a.st.ListKillCriteria(ideaID)
}

// ListReviewsDue runs the on-launch sweep: pending kill criteria due at or
// before now, shaped with days-overdue for the Reviews Due panel.
func (a *API) ListReviewsDue() ([]pipeline.DueReviewView, error) {
	now := time.Now().UnixMilli()
	rows, err := a.st.ListDueKillCriteria(now)
	if err != nil {
		return nil, err
	}
	return pipeline.ShapeDueReviews(rows, now), nil
}
